### PR TITLE
Add support for loading KTX Textures

### PR DESCRIPTION
**Pull request by achalpandeyy**
Currently only works for 2D textures.

TODO: Multi Layered Textures (Texture Arrays) & Cube Maps.

Includes a KTX texture file for testing purposes.

### CODE REVIEW

**Comment by devshgraphicsprogramming** [comment]
Do multilayered textures and cubemaps then we merge ;)

File: CMakeLists.txt [hunk 1]
```diff
@@ -183,8 +183,20 @@ option(IRR_BUILD_BULLET "Enable Bullet Physics building and integration?" OFF)
...
 option(IRR_BUILD_DOCS "Enable building documentation?" ON)
 
+option(IRR_COMPILE_WITH_GLI "Enable building with GLI?" ON)
+
+option(IRR_COMPILE_WITH_KTX_LOADER "Enable building with KTX LOADER?" ON)
+
 set(THIRD_PARTY_SOURCE_DIR "${PROJECT_SOURCE_DIR}/3rdparty")
 
+if(IRR_COMPILE_WITH_GLI)
+	add_definitions(-D_IRR_COMPILE_WITH_GLI_)
```
Comment (devshgraphicsprogramming) [inline CMakeLists.txt:193]: not like this, look at https://github.com/buildaworldnet/IrrlichtBAW/blob/master/include/irr/config/BuildConfigOptions.h.in

File: src/irr/asset/CImageLoaderKTX.cpp [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+SOFTWARE.
+*/
+
+#include "CImageLoaderKTX.h"
+
+#ifdef _IRR_COMPILE_WITH_KTX_LOADER_
+
+#ifdef _IRR_COMPILE_WITH_GLI_
+#include "gli/gli.hpp"
```
Comment (devshgraphicsprogramming) [inline src/irr/asset/CImageLoaderKTX.cpp:26]:
```cpp
#else
#error "Need GLI"
```

File: src/irr/asset/CImageLoaderKTX.cpp [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+        if (buffer[i] != signature[i])
+            return false;
+    }
+    return true;
+}
+
+//! returns true if the file maybe is able to be loaded by this class
+bool CImageLoaderKTX::isALoadableFileFormat(io::IReadFile* _file) const
+{
+#ifdef _IRR_COMPILE_WITH_GLI_
```
Comment (devshgraphicsprogramming) [inline src/irr/asset/CImageLoaderKTX.cpp:57]: if you did https://github.com/buildaworldnet/IrrlichtBAW/pull/381/files?file-filters%5B%5D=.cpp&file-filters%5B%5D=.h&file-filters%5B%5D=.ktx&file-filters%5B%5D=.txt&file-filters%5B%5D=No+extension#r336541245

then we wouldn't need this `#ifdef` clutter

File: src/irr/asset/CImageLoaderKTX.cpp [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+        os::Printer::log("Failed to load texture at path ", filename, ELL_ERROR);
+        return {};
+    }
+
+    // Instantiate translation class to convert GLI enums into OpenGL values
+    gli::gl GL(gli::gl::PROFILE_GL33);
+    const gli::gl::format textureFormat = GL.translate(texture.format(), texture.swizzles());
+    GLenum textureTarget = GL.translate(texture.target());
+
+    uint32_t nullOffset[3] = { 0, 0, 0 };
```
Comment (devshgraphicsprogramming) [inline src/irr/asset/CImageLoaderKTX.cpp:123]: if you want to support 1d,2d array textures or cubemaps you need to call this something else, and touch index 1 or 2 so you have the correct 3d extent

Comment (devshgraphicsprogramming) [reply src/irr/asset/CImageLoaderKTX.cpp:123]: I suggest to rename it to `minOffset`

File: src/irr/asset/CImageLoaderKTX.cpp [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+    for (unsigned int mipLevel = 0; mipLevel < texture.levels(); mipLevel++)
+    {
+        uint32_t imageSize[3] = { texture.extent(mipLevel)[0], texture.extent(mipLevel)[1], texture.extent(mipLevel)[2] };
+        for (unsigned int face = 0; face < texture.faces(); face++)
+        {
+            asset::CImageData* image = nullptr;
```
Comment (devshgraphicsprogramming) [inline src/irr/asset/CImageLoaderKTX.cpp:129]: you need to do
```cpp
minOffset[is1D ? 1:2] = face;
```

then everything shall work

**Comment by devshgraphicsprogramming** [comment]
@achalpandeyy how's work going?

**Comment by achalpandeyy** [comment]
Thanks for the review!
Going great but incredibly busy, hopefully I will be able to make changes by the end of this week

File: src/irr/asset/CImageLoaderKTX.cpp [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+    if (_file->read(buffer, 12) != 12)
+    {
+        _file->seek(prevPos);
+        return false;
+    }
+
+    _file->seek(prevPos);
+
+    // Check if it really is a KTX _file
+    return checkKTXSignature(buffer);
```
Comment (devshgraphicsprogramming) [inline src/irr/asset/CImageLoaderKTX.cpp:75]: gli loads both DDS and KTX, you could nuke our DDS loader and have this class handle both KTX and DDS.

File: src/irr/asset/CImageLoaderKTX.cpp [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+
+    // Check if it really is a KTX _file
+    if (!checkKTXSignature(buffer))
+    {
+        os::Printer::log("LOAD KTX: not really a ktx\n", _file->getFileName().c_str(), ELL_ERROR);
+        return {};
+    }
+
+    const char* filename = _file->getFileName().c_str();
+    gli::texture texture = gli::load_ktx(filename);
```
Comment (devshgraphicsprogramming) [inline src/irr/asset/CImageLoaderKTX.cpp:111]: ktx and dds would be nice.

File: src/irr/asset/CImageLoaderKTX.h [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+    Surface Loader for KTX images
+*/
+class CImageLoaderKTX : public asset::IAssetLoader
+{
+public:
+    virtual bool isALoadableFileFormat(io::IReadFile* _file) const override;
+
+    virtual const char** getAssociatedFileExtensions() const override
+    {
+        static const char* ext[]{ "ktx", nullptr };
```
Comment (devshgraphicsprogramming) [inline src/irr/asset/CImageLoaderKTX.h:46]: would be good to load dds as well since GLI does both (and you could get rid of our awful DDS loader).

**Comment by devshgraphicsprogramming** [comment]
Either port this loader to `shader_pipeline` and do a PR from that branch, or let @AnastaZIuk take over.

**Comment by achalpandeyy** [comment]
Sure, @AnastaZIuk can take over.

**Comment by devshgraphicsprogramming** [comment]
superseded by #402

**Pull request closed without merging**